### PR TITLE
Wat.068

### PR DESCRIPTION
**Pull request by rachelthoms**
## Checklist for Reviewing a Pre-Processing Script

- [ ] Does the python script contain the following 4 sections: Download data and save to your data directory, Process data, Upload processed data to Carto/Upload processed data to Google Earth Engine, Upload original data and processed data to Amazon S3 storage?
- [ ] Does the script have the standardized variable names for: dataset_name, raw_data_file, processed_data_file?
- [ ] Does the script create and use a 'data' directory?
- [ ] Does the script use a python module to automatically download the data? If this is not possible, are there explicit instructions for how to download the data (step by step with instructions about every input parameter and button to click to find the exact data) and does it use shutil to move the data from 'Downloads' into the data directory?
- [ ] Is the script automated as much as possible to minimize rewriting code the next time the dataset updates (ex 1: can you automatically pull out column names instead of typing them yourself? ex 2: if you are dropping columns with no data, did you use pandas to find the nodata columns instead of dropping the column by name?)
- [ ] Are there comments on almost every line of code to explicitly state what is being done and why?
- [ ] Are you uploading to the correct AWS location?
- [ ] For GEE scripts, did you explicitly define the band manifest with a pyramiding policy so that we can easily change it later if we need to?
- [ ] Does the README contain all relevant links?
- [ ] Does the README state the original file type?
- [ ] Does the README list all of the processing steps that were taken in the script?
- [ ] For netcdfs, does the README state which variables were pulled from the netcdf?
- [ ] Did you use the util functions whenever possible?
- [ ] Have you checked the processed data file on your computer to make sure it matches the data you uploaded to Carto? (spaces or symbol names in column titles often get changed in the upload process-please change these before uploading so that the backed up processed data matches the data on Carto)
- [ ] Does the folder name and python script file name match the dataset_name variable in the script? Are they all lowercase? Does the processing script end in '_processing.py'? (this part is often forgotten)

### CODE REVIEW

**Review by adawyj97 — CHANGES_REQUESTED** [review]

File: wat_068_rw0_watersheds/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+This file describes the data pre-processing for the [HydroBASINS](https://www.hydrosheds.org/page/hydrobasins) dataset for [display on Resource Watch](https://resourcewatch.org/data/explore/TOBEDETERMINED).
+
+The source provided this dataset as a series of zipped shapefiles accessed through its [website](https://www.dropbox.com/sh/hmpwobbz9qixxpe/AACPCyoHHAQUt_HNdIbWOFF4a/HydroBASINS/standard?dl=0&subfolder_nav_tracking=1). 
+
+Below, we describe the steps used to reformat the shapefiles to upload them to Carto.
+
+1. For each relevant basin level, read in the shapefiles for the nine regional tiles as a geopandas dataframe.
```
[the remark this answers is not in the document]
Comment (rachelthoms) [reply wat_068_rw0_watersheds/README.md:11]: let me know if the recent push clarifies these points

File: wat_068_rw0_watersheds/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+## Watersheds Data Pre-Processing
+This file describes the data pre-processing for the [HydroBASINS](https://www.hydrosheds.org/page/hydrobasins) dataset for [display on Resource Watch](https://resourcewatch.org/data/explore/ab6216ee-9a2b-412f-b538-8644d5834c7a).
+
+The source provided data as regional tiles in individual polygon shapefiles, one for each
+region and each Pfafstetter level. 
+
+Due to the resolution of the sub-basins delinations, the breakdowns provided by levels 3-8 are most relevant for analysis, and were thus selected for display on Resource Watch.
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/README.md:7]:
```suggestion
Due to the resolution of the sub-basins delineations, the breakdowns provided by levels 3-8 are most relevant for analysis, and were thus selected for display on Resource Watch. 
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        zip.extractall(data_dir)
+
+'''
+Process the data and upload to carto 
+'''
+
+# Define object that stores which basin levels will be processed. There are 12 basin levels for each region.
+include_levels = [False, False, True, True, True, True, True, True, False, False, False, False]
+
+
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:85]:
```suggestion
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    
+    # Test if the basin level should be included, using the index
+    if not include_levels[i]:
+        continue
+   
+    else:
+        # create an empty list to hold the gdfs to be merged (1 per region)
+        gdf_list = []
+       
+        # create a str of the basin level padded by zeros, using the index
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:97]:
```suggestion
        # create a string of the basin level padded by zeros, using the index
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        n = str(i+1).zfill(2)
+
+        # iterate through the list of extracted file names using a for loop
+        for name in raw_namelist:
+            
+            # read shapefiles that include the corresponding basin level as a gdf
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:103]:
```suggestion
            # read shapefiles that include the corresponding basin level as a geopandas dataframe
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        processed_data_file = os.path.join(data_dir, dataset_name+ '_lev'+ n + '_edit.shp')
+
+        # save processed dataset to shapefile
+        logger.debug('Creating shapefile')
+        out_gdf.to_file(processed_data_file,driver='ESRI Shapefile')
+    
+        # Create a string of the file name of the processed data files, excluding the file extension
+        out_str = processed_data_file[:-3]+'*'
+        
+        # Create a list of the shapefile components by grabbing files that match the str
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:130]:
```suggestion
        # Create a list of the shapefile components by grabbing files that match the string
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                gdf = gpd.read_file("data/" + name)
+                
+                # Add the gdf to the list
+                gdf_list.append(gdf) 
+                
+                # Add level column, pulled from the index
+                level = i+1
+                gdf['level'] = level
+        
+        # merge gdfs for basin level
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:114]:
```suggestion
        # merge geopandas dataframes for basin level
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+# Read included basin-level shapefiles into Python using a for loop
+for i in range(12):
+    
+    # Test if the basin level should be included, using the index
+    if not include_levels[i]:
+        continue
+   
+    else:
+        # create an empty list to hold the gdfs to be merged (1 per region)
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:94]:
```suggestion
        # create an empty list to hold the geopandas dataframes to be merged (1 per region)
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py (hunk 5 above)
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:103]:
```suggestion
            # read shapefiles that include the corresponding basin level as a geopandas dataframe
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        n = str(i+1).zfill(2)
+
+        # iterate through the list of extracted file names using a for loop
+        for name in raw_namelist:
+            
+            # read shapefiles that include the corresponding basin level as a gdf
+            if n in name and name.endswith('.shp'):
+                gdf = gpd.read_file("data/" + name)
+                
+                # Add the gdf to the list
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:107]:
```suggestion
                # Add the geopandas dataframe to the list
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        with ZipFile(out_zip_dir,'w') as zip:
+            for file in out_filelist:
+                zip.write(file, os.path.basename(file))
+                zipfile_list.append(file)
+
+        # Upload zipfile to carto
+        logger.info('Uploading processed data for ' + processed_data_file +' to Carto.')
+        util_carto.upload_to_carto(out_zip_dir, 'LINK')
+        
+
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:148]:
```suggestion
```

File: wat_068_rw0_watersheds/README.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+The source provided data as regional tiles in individual polygon shapefiles, one for each
+region and each Pfafstetter level. 
+
+Due to the resolution of the sub-basins delinations, the breakdowns provided by levels 3-8 are most relevant for analysis, and were thus selected for display on Resource Watch. 
+
+Below, we describe the steps used to reformat the shapefiles to upload them to Carto.
+
+1. For each relevant basin level, read in the shapefiles for the nine regional tiles as a geopandas dataframe.
+2. Combine shapefiles for each basin level into one shapefile that includes all regional tiles, with a column for 'level' to indicate which level the file represents.
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/README.md:12]:
```suggestion
2. Convert column names to lowercase to match Carto column name requirements.
3. Combine shapefiles for each basin level into one shapefile that includes all regional tiles, with a column for 'level' to indicate which level the file represents.
```

**Review by adawyj97 — CHANGES_REQUESTED** [review]

File: wat_068_rw0_watersheds/README.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+region and each Pfafstetter level. 
+
+Due to the resolution of the sub-basins delineations, the breakdowns provided by levels 3-8 are most relevant for analysis, and were thus selected for display on Resource Watch. 
+
+Below, we describe the steps used to combine the regional shapefiles by level and insert these into a single carto table.
+
+For each relevant basin level:
+1. Read in the shapefiles for the nine regional tiles as a geopandas dataframe.
+2. Combine shapefiles for all regional tiles at the given level into one shapefile, with a column for 'level' to indicate which level the file represents.
+3. Upload the combined shapefile to Carto.
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/README.md:14]:
```suggestion
For each relevant basin level:
1. Read in the shapefiles for the nine regional tiles as a geopandas dataframe.
2. Convert the column names to lowercase to match Carto column name requirements.
3. Combine shapefiles for all regional tiles at the given level into one shapefile, with a column for 'level' to indicate which level the file represents.
4. Upload the combined shapefile to Carto.
```

File: wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+# name of table on Carto where you want to upload data
+# this should be a table name that is not currently in use
+dataset_name = 'wat_068_rw0_watersheds' #check
+
+logger.info('Executing script for dataset: ' + dataset_name)
+# create a new sub-directory within your specified dir called 'data'
+# within this directory, create files to store raw and processed data
+data_dir = util_files.prep_dirs(dataset_name)
+
+
```
Comment (adawyj97) [inline wat_068_rw0_watersheds/wat_068_rw0_watersheds_processing.py:39]:
```suggestion

```

**Review by adawyj97 — CHANGES_REQUESTED** [review]

**Pull request merged**